rfc/user-secrets.md

```diff
@@ -0,0 +1,186 @@
+
+```json
+{
+  "type": "opaque",
+  "roles": ["admin", "sre"],
+  "data": {
+    "kubeconfig": "YXBpVmVyc2lvbjogdjEKY2x1c3RlcnM6..."
+  }
+}
+```
+
+Then using this secret in a new Clouddriver account:
+
+```json
+{
+  "type": "kubernetes",
+  "name": "eks",
+  "permissions": {
+    "READ": ["admin", "sre", "dev"],
+    "WRITE": ["admin", "sre"]
+  },
+  "context": "eks",
+  "namespaces": ["default", "sandbox"],
+  "kubeconfigContents": "secret://secrets-manager?r=us-east-1&s=arn:aws:secretsmanager:us-east-1:123455432100:secret:eks-sa&e=json&k=kubeconfig"
+}
+```
+
+Note that secrets can be referred to by just the secret name and not the full ARN when the secret is stored in the same Spinnaker workload AWS account.
+However, it is expected that in a multi-tenant environment, different tenants may have different AWS accounts, thus the tenants are responsible for exporting their own secrets to Spinnaker through IAM policies.
+Similar authorization strategies should be used in other external secrets managers.
+
+### Dependencies
+
+This feature updates `kork-secrets` and `kork-secrets-aws`.
+Existing Jackson Databind dependencies are used for JSON, YAML, and CBOR support.
+
+## Drawbacks
+
+User secrets requires two levels of authorization: authorization for the Spinnaker service(s) to read secrets from a secrets manager service, and authorization for users inside the Spinnaker service(s) to use these secrets.
+Existing audit logs of secrets access in the external secrets manager may lose context on how secrets are used without aggregating audit logs from Spinnaker.
+Secrets rotation scripts need to be updated to respect the format in use.
+
+## Prior Art and Alternatives
+
+Kork provides the `EncryptedSecret` class and corresponding URI formats for injecting secrets via Halyard and some limited support in Spinnaker services to use the same API.
+This API is designed for single tenants and provides no access controls.
+
+Dedicated secrets managers such as [Vault](https://www.vaultproject.io) provide much more sophisticated secrets management and delivery methods, though it is a complex system to manage in addition to Spinnaker and the cloud infrastructure where it runs.
+While Vault could become another `SecretEngine` provider in Kork, requiring Vault when some users may already be using cloud-specific secrets managers or other secrets manager software would be too inflexible.
+
+It may be possible to store the Spinnaker authorization data for secrets in secrets metadata APIs, though the implementation details on how each secrets manager exposes metadata varies quite a bit.
+Since these access controls are not particularly sensitive, they don't _need_ to be encrypted with the secret, though this is the most direct way to deliver said access controls such that Fiat can understand them.
+
+This `UserSecretManager` API would replace the tentative use of `SecretManager` in the Clouddriver Account Management API.
```

Comment:
Is there a plan to replace all the existing secret engine classes with this newer api/format structure? E.g. treat ALL secrets as user secrets?

Reply:
User secrets are more of an extension to external secrets. Maybe they can be unified somehow?